### PR TITLE
Cookbook for adding tabs

### PR DESCRIPTION
**Pull request by jordygroote**
#294 ## What's in this PR?

As discussed with @alexander-schranz.
I created a cookbook for adding tabs to Sulu admin. I think this could be very useful for more people :).

Please review it, if some things are wrong i will change it.

PS (Don't know if i created this pull request the right way.)

### CODE REVIEW

**Comment by jordygroote** [comment]
2 things i will fix:

- I forgot i had a TODO in the commit, will have to create a link to an internal document
- `The CI is failing thats mostly I think in your case that the headline (====, -----) need to be the same length as the text of the headline.`

**Comment by jordygroote** [comment]
@danrot  @alexander-schranz I fixed the 2 to-do's. This is now ready for review. Hope you can check it out :)

File: cookbook/add-tabs.rst [hunk 1]
```diff
@@ -0,0 +1,300 @@
+Adding tabs to Sulu admin interface
```
Comment (danrot) [inline cookbook/add-tabs.rst:1]: In order to better match other headings I would rephrase it to "Adding tabs to Sulu's Admin UI"

File: cookbook/add-tabs.rst [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+===================================
+
+.. note:: It is recommended to read `Extend admin Ui`_ beforehand to get a better understandig of how Sulu admin classes work.
```
Comment (danrot) [inline cookbook/add-tabs.rst:4]: The link is not working, and could you please change the wording to "Extend Admin UI", to match the case int he other page?

File: cookbook/add-tabs.rst [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+
+
```
Comment (danrot) [inline cookbook/add-tabs.rst:6]: Only one new line

File: cookbook/add-tabs.rst [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+This tutorial will walk you through the process of adding extra tabs to the user interface of Sulu.
```
Comment (danrot) [inline cookbook/add-tabs.rst:7]: "of adding an extra tab" (you are only adding one tab in this entry, right?)

File: cookbook/add-tabs.rst [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+
+This could be useful in a few situations, a few examples below:
```
Comment (danrot) [inline cookbook/add-tabs.rst:9]: The examples are almost the same... Could we maybe write something like this without a listing?

"This could be useful in many different situations, like e.g. adding an extra tab for social media information to pages or contacts."

File: cookbook/add-tabs.rst [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+
+- Saving extra data for pages
+- Saving extra data in webspaces
+- Saving extra data for users
+
+In this example we'll be adding a "Socials" tab to the PageAdmin of Sulu.
```
Comment (danrot) [inline cookbook/add-tabs.rst:15]: "to the page form of Sulu" - Otherwise it is not really clear, the Admin class can add to a lot of things.

File: cookbook/add-tabs.rst [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+
+.. figure:: ../img/socials-tab.png
```
Comment (danrot) [inline cookbook/add-tabs.rst:17]: Can you make a screen shot of the full browser window? Otherwise it looks a little bit weird, because the toolbar is only partly visible. I think you should use a tool like "FireShot" and make the window a bit smaller.

File: cookbook/add-tabs.rst [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+
+Create the form for this view
+-----------------------------
+You have to create the form that is rendered in this view. For this we create an XML file like your used to.
```
Comment (danrot) [inline cookbook/add-tabs.rst:21]: Add a new line between headings and texts.

Comment (danrot) [inline cookbook/add-tabs.rst:21]: Would omit the second sentence.

File: cookbook/add-tabs.rst [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+Create your form: ``config/forms/page_socials.xml``. Notice how we use slashes in the names of the properties, this will be of later use.
```
Comment (danrot) [inline cookbook/add-tabs.rst:22]: "Therefore create a form at ``config/forms/page_socials.xml``:"

The notice could be an actual note using the `.. note::` syntax. Also explain what it is doing (returning the values in the given hierarchy) instead of just saying that it will be of use later.

File: cookbook/add-tabs.rst [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+    <form xmlns="http://schemas.sulu.io/template/template"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="http://schemas.sulu.io/template/template http://schemas.sulu.io/template/form-1.0.xsd"
+    >
+        <key>page_socials</key>
+
+        <properties>
+            <section name="twitter">
+                <meta>
+                    <title>Twitter</title>
```
Comment (danrot) [inline cookbook/add-tabs.rst:36]: This should cause warnings in the UI, because if you don't pass a language to the `title` tag it will be tried to be used as a translation key, and this translation key does not exist. We could work around that by writing it the following way instead:

```xml
<title lang="en">Twitter</title>
```

File: cookbook/add-tabs.rst [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+                </meta>
+                <properties>
+                    <property name="ext/social/twitter_title" type="text_line">
+                        <meta>
+                            <title>Twitter title</title>
```
Comment (danrot) [inline cookbook/add-tabs.rst:41]: Same as above

File: cookbook/add-tabs.rst [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+                        </meta>
+                    </property>
+                    <property name="ext/social/twitter_description" type="text_line">
+                        <meta>
+                            <title>Twitter description</title>
```
Comment (danrot) [inline cookbook/add-tabs.rst:46]: Same as above

File: cookbook/add-tabs.rst [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+                        </meta>
+                    </property>
+                    <property name="ext/social/twitter_image" type="single_media_selection">
+                        <meta>
+                            <title>Twitter image</title>
```
Comment (danrot) [inline cookbook/add-tabs.rst:51]: Same as above

File: cookbook/add-tabs.rst [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+                            </param>
+                        </params>
+                    </property>
+                </properties>
+            </section>
+        </properties>
+    </form>
+
+
+
```
Comment (danrot) [inline cookbook/add-tabs.rst:66]: Only one new line

File: cookbook/add-tabs.rst [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+Registering your own admin class
+--------------------------------
+
+We'll create a class for example ``src/Admin/PageAdminExtension`` which extends ``Sulu\Bundle\AdminBundle\Admin``.
+
+In the admin we need to implement configureViews. Also for our example we use the `webspace_manager` in combination with `security_checker` to check if the logged in user has permission to edit pages.
```
Comment (danrot) [inline cookbook/add-tabs.rst:72]: Put code reference in two backticks (talking about `configureViews` here)

Comment (danrot) [inline cookbook/add-tabs.rst:72]: And "In the `Admin` class".

Comment (danrot) [inline cookbook/add-tabs.rst:72]: "For our example we need the ..."

File: cookbook/add-tabs.rst [hunk 16]
```diff
@@ -0,0 +1,300 @@
...
+
+Also important, notice how we set the parent to be `PageAdmin::EDIT_FORM_VIEW`. This adds our view to the `PageAdmin::EDIT_FORM_VIEW` view.
```
Comment (danrot) [inline cookbook/add-tabs.rst:74]: "Notice the `setParent` call adding the view as a child to `PageAdmin::EDIT_FORM_VIEW`. This will result in a new tab in the edit form for pages.

File: cookbook/add-tabs.rst [hunk 17]
```diff
@@ -0,0 +1,300 @@
...
+
+The setFormKey is a reference to the <key> in the created form from the above section.
```
Comment (danrot) [inline cookbook/add-tabs.rst:76]: "`setFormKey` takes a string reference, which should be the same as the `key` tag in the above form XML."

File: cookbook/add-tabs.rst [hunk 18]
```diff
@@ -0,0 +1,300 @@
...
+
+.. code-block:: php
+
+    class PageAdminExtension extends Admin
```
Comment (danrot) [inline cookbook/add-tabs.rst:80]: I think I would call this `SocialAdmin` or something like this... But let me talk to my colleagues about that first.

File: cookbook/add-tabs.rst [hunk 19]
```diff
@@ -0,0 +1,300 @@
...
+                        ->disablePreviewWebspaceChooser()
+                        ->setResourceKey('pages')
+                        ->setFormKey('page_socials')
+                        ->setTabTitle('Socials')
+                        ->setTabPriority(511)
+                        ->addToolbarActions($formToolbarActionsWithoutType)
+                        ->addRouterAttributesToFormRequest($routerAttributesToFormRequest)
+                        ->setPreviewCondition($previewCondition)
+                        ->setTitleVisible(true)
+                        ->setTabOrder(1025)
```
Comment (danrot) [inline cookbook/add-tabs.rst:131]: Actually we tend to use binary numbers here, so that you could add more stuff in between later. So in that case it would be `2048`.

File: cookbook/add-tabs.rst [hunk 20]
```diff
@@ -0,0 +1,300 @@
...
+
+            return false;
+        }
+    }
+
+We can register this class as a service and give it a tag of ``sulu.admin``, then it wil be picked up by Sulu.
+
+.. code-block:: yaml
+
+    my_own.admin:
```
Comment (danrot) [inline cookbook/add-tabs.rst:158]: I would call that service `app.social_admin`

File: cookbook/add-tabs.rst [hunk 21]
```diff
@@ -0,0 +1,300 @@
...
+        class: App\Admin\PageAdminExtension
+        arguments:
+            - '@Sulu\Bundle\AdminBundle\Admin\View\ViewBuilderFactoryInterface'
+            - '@sulu_core.webspace.webspace_manager'
+            - '@sulu_security.security_checker'
+        tags:
+            - { name: 'sulu.admin'}
+            - { name: 'sulu.context', context: 'admin' }
+
+When you debug the container right now your should see your own admin show up.
```
Comment (danrot) [inline cookbook/add-tabs.rst:168]: "you should see your own `Admin` class show up."

File: cookbook/add-tabs.rst [hunk 22]
```diff
@@ -0,0 +1,300 @@
...
+        sulu_tag.admin           Sulu\Bundle\TagBundle\Admin\TagAdmin                
+        sulu_page.admin          Sulu\Bundle\PageBundle\Admin\PageAdmin              
+        sulu_snippet.admin       Sulu\Bundle\SnippetBundle\Admin\SnippetAdmin        
+        sulu_category.admin      Sulu\Bundle\CategoryBundle\Admin\CategoryAdmin      
+        sulu_security.admin      Sulu\Bundle\SecurityBundle\Admin\SecurityAdmin      
+        sulu_media.admin         Sulu\Bundle\MediaBundle\Admin\MediaAdmin            
+        sulu_search.admin        Sulu\Bundle\SearchBundle\Admin\SearchAdmin          
+        my_own.admin             App\Admin\PageAdminExtension  
+
+You should now see the tab show up but it's not working just now.
```
Comment (danrot) [inline cookbook/add-tabs.rst:188]: "not working yet"

File: cookbook/add-tabs.rst [hunk 23]
```diff
@@ -0,0 +1,300 @@
...
+
+
+
```
Comment (danrot) [inline cookbook/add-tabs.rst:191]: Only one new line

File: cookbook/add-tabs.rst [hunk 24]
```diff
@@ -0,0 +1,300 @@
...
+Registering your structure extension 
+------------------------------------
+
+For this all to work we need to make a structure extension which handles the saving and loading of the data.
```
Comment (danrot) [inline cookbook/add-tabs.rst:195]: "For this to work we need to build a structure extension handling the storage of additional data for pages."

File: cookbook/add-tabs.rst [hunk 25]
```diff
@@ -0,0 +1,300 @@
...
+
+.. code-block :: php
+    
+    <?php
+
+    /**
+    * extends structure with social content.
+    */
```
Comment (danrot) [inline cookbook/add-tabs.rst:203]: Remove that comment, as it does not really add any value.

File: cookbook/add-tabs.rst [hunk 26]
```diff
@@ -0,0 +1,300 @@
...
+        {
+            if ($data['twitter_image']) {
+                $data['twitter_image'] = json_encode($data['twitter_image']);
+            }
+
+            return $data;
+        }
+    }
+
+Also this should be registered as a service with the tag `sulu.structure.extension`.
```
Comment (danrot) [inline cookbook/add-tabs.rst:288]: "This service needs to be registered as a service with the tag `sulu.structure.extension`."

File: cookbook/add-tabs.rst [hunk 27]
```diff
@@ -0,0 +1,300 @@
...
+
+.. code-block :: yaml
+
+    my_own.extension.social:
```
Comment (danrot) [inline cookbook/add-tabs.rst:292]: `app.social_structure_extension`

File: cookbook/add-tabs.rst [hunk 28]
```diff
@@ -0,0 +1,300 @@
...
+        class: App\Structure\SocialStructureExtension
+        tags: { name: 'sulu.structure.extension' }
+
+Because we used slashes in the property names in the form. Our data is automaticly mapped to the extension data given in the view.
```
Comment (danrot) [inline cookbook/add-tabs.rst:296]: That could probably be added in the note I've mentioned above, then we can remove it down here. Also, it does still not really explain what the slashes are doing.

File: cookbook/add-tabs.rst [hunk 29]
```diff
@@ -0,0 +1,300 @@
...
+
+.. code-block :: yaml
+
+    my_own.extension.social:
+        class: App\Structure\SocialStructureExtension
+        tags: { name: 'sulu.structure.extension' }
+
+Because we used slashes in the property names in the form. Our data is automaticly mapped to the extension data given in the view. 
+
+.. figure:: ../img/socials-tab-dump.png
```
Comment (danrot) [inline cookbook/add-tabs.rst:298]: Wouldn't add that as a screenshot... Maybe we can add a table showing how to access which value. The thing is that the screenshot is not searchable.

**Review by danrot — CHANGES_REQUESTED** [review]
Maybe we should rename the file from `add-tabs` to `add-page-tabs`, since some of the information is only valid for pages.

**Comment by danrot** [comment]
Oh, totally forgot! Thanks for your effort! :heart:

**Comment by jordygroote** [comment]
Haha sure, no problem. 
I'll see if i have time this week to resolve the feedback :)

File: cookbook/add-tabs.rst (hunk 26 above)
Comment (jordygroote) [reply cookbook/add-tabs.rst:288]: "This class needs to be registered as a service with the tag sulu.structure.extension." 
Right?

File: cookbook/add-tabs.rst (hunk 28 above)
Comment (jordygroote) [reply cookbook/add-tabs.rst:296]: removed here

File: cookbook/add-tabs.rst (hunk 29 above)
Comment (jordygroote) [reply cookbook/add-tabs.rst]: I removed the screenshot. Not sure what you mean by the table.

File: cookbook/add-tabs.rst (hunk 2 above)
Comment (jordygroote) [reply cookbook/add-tabs.rst]: Not sure why the link is not working. Can you have a look?

**Comment by jordygroote** [comment]
I resolved most of your feedback. 
Placed some comments under your conversation bubbles.

File: cookbook/add-tabs.rst (hunk 18 above)
Comment (nnatter) [reply cookbook/add-tabs.rst:77]: `SocialAdmin` sounds good to me 🙂

File: cookbook/add-tabs.rst [hunk 30]
```diff
@@ -0,0 +1,289 @@
+Adding tabs to Sulu's Admin UI
+==============================
+
+.. note:: It is recommended to read `Extend Admin UI`_ beforehand to get a better understandig of how Sulu admin classes work.
+
+This tutorial will walk you through the process of adding an extra tab.
```
Comment (nnatter) [inline cookbook/add-tabs.rst:6]:
```suggestion
This tutorial will walk you through the process of adding an extra tab to the administration interface.
```

File: cookbook/add-tabs.rst [hunk 31]
```diff
@@ -0,0 +1,289 @@
...
+                </properties>
+            </section>
+        </properties>
+    </form>
+
+
+Registering your own admin class
+--------------------------------
+
+We'll create a class for example ``src/Admin/PageAdminExtension`` which extends ``Sulu\Bundle\AdminBundle\Admin``.
```
Comment (nnatter) [inline cookbook/add-tabs.rst:67]:
```suggestion
We'll create a class for example ``src/Admin/SocialAdmin`` which extends ``Sulu\Bundle\AdminBundle\Admin``.
```

File: cookbook/add-tabs.rst [hunk 32]
```diff
@@ -0,0 +1,289 @@
...
+        sulu_tag.admin           Sulu\Bundle\TagBundle\Admin\TagAdmin                
+        sulu_page.admin          Sulu\Bundle\PageBundle\Admin\PageAdmin              
+        sulu_snippet.admin       Sulu\Bundle\SnippetBundle\Admin\SnippetAdmin        
+        sulu_category.admin      Sulu\Bundle\CategoryBundle\Admin\CategoryAdmin      
+        sulu_security.admin      Sulu\Bundle\SecurityBundle\Admin\SecurityAdmin      
+        sulu_media.admin         Sulu\Bundle\MediaBundle\Admin\MediaAdmin            
+        sulu_search.admin        Sulu\Bundle\SearchBundle\Admin\SearchAdmin          
+        my_own.admin             App\Admin\PageAdminExtension  
+
+You should now see the tab show up but it's not working yet.
```
Comment (nnatter) [inline cookbook/add-tabs.rst:185]:
```suggestion
You should now see the tab in the administration interface, but the data of the form is not saved yet.
```

File: cookbook/add-tabs.rst [hunk 33]
```diff
@@ -0,0 +1,289 @@
...
+                            </param>
+                        </params>
+                    </property>
+                </properties>
+            </section>
+        </properties>
+    </form>
+
+
+Registering your own admin class
```
Comment (nnatter) [inline cookbook/add-tabs.rst:64]:
```suggestion
Register the view in your admin class
```

File: cookbook/add-tabs.rst [hunk 34]
```diff
@@ -0,0 +1,289 @@
...
+        sulu_category.admin      Sulu\Bundle\CategoryBundle\Admin\CategoryAdmin      
+        sulu_security.admin      Sulu\Bundle\SecurityBundle\Admin\SecurityAdmin      
+        sulu_media.admin         Sulu\Bundle\MediaBundle\Admin\MediaAdmin            
+        sulu_search.admin        Sulu\Bundle\SearchBundle\Admin\SearchAdmin          
+        my_own.admin             App\Admin\PageAdminExtension  
+
+You should now see the tab show up but it's not working yet.
+
+
+Registering your structure extension
```
Comment (nnatter) [inline cookbook/add-tabs.rst:188]: This is quite page specific. Should we use some more generic title like `Persist the data of the form`?
 @danrot

File: cookbook/add-tabs.rst [hunk 35]
```diff
@@ -0,0 +1,289 @@
...
+        sulu_search.admin        Sulu\Bundle\SearchBundle\Admin\SearchAdmin          
+        my_own.admin             App\Admin\PageAdminExtension  
+
+You should now see the tab show up but it's not working yet.
+
+
+Registering your structure extension 
+------------------------------------
+
+For this to work we need to build a structure extension handling the storage of additional data for pages.
```
Comment (nnatter) [inline cookbook/add-tabs.rst:191]: What do you think about something like:
```
In the final step, we need to persist the data of the added tab. 
In the case of the pages, we can utilize the existing pages API by registering 
a new `StructureExtension`. 
In other cases, we would need to implement our own API endpoint for the tab 
as shown in the `Extend Admin UI`_ chapter.
```

**Review by nnatter — COMMENTED** [review]
Thanks a lot for you work! I think it would be nice to target the `2.0` branch 🙂

**Review by danrot — CHANGES_REQUESTED** [review]
And can you please make the screenshot without the window borders? Extensions like [FireShot](https://chrome.google.com/webstore/detail/take-webpage-screenshots/mcbpblocgmgfnpjjppndjkmgjaogfceg) would help with that.

File: cookbook/add-tabs.rst [hunk 36]
```diff
@@ -0,0 +1,289 @@
+Adding tabs to Sulu's Admin UI
+==============================
+
+.. note:: It is recommended to read `Extend Admin UI`_ beforehand to get a better understandig of how Sulu admin classes work.
```
[the remark this answers is not in the document]
Comment (jordygroote) [reply cookbook/add-tabs.rst:4]: OK will check.

File: cookbook/add-tabs.rst [hunk 37]
```diff
@@ -0,0 +1,289 @@
...
+
+.. figure:: ../img/socials-tab.png
+
+Create the form for this view
+-----------------------------
+
+You have to create the form that is rendered in this view.
+Therefore create a form at ``config/forms/page_socials.xml``. 
+
+.. note:: Notice how we use slashes in the names of the properties, this returns the values in the given hierarchy.
```
[the remark this answers is not in the document]
Comment (jordygroote) [reply cookbook/add-tabs.rst:20]: OK.

File: cookbook/add-tabs.rst [hunk 38]
```diff
@@ -0,0 +1,289 @@
...
+        </properties>
+    </form>
+
+
+Register the view in your admin class
+--------------------------------
+
+We'll create a class for example ``src/Admin/SocialAdmin`` which extends ``Sulu\Bundle\AdminBundle\Admin``.
+
+In the ``Admin`` we need to implement ``configureViews``. For our example we need the `webspace_manager` in combination with `security_checker` to check if the logged in user has permission to edit pages.
```
[the remark this answers is not in the document]
Comment (jordygroote) [reply cookbook/add-tabs.rst:69]: Ah OK.

File: cookbook/add-tabs.rst [hunk 39]
```diff
@@ -0,0 +1,289 @@
...
+                        </params>
+                    </property>
+                </properties>
+            </section>
+        </properties>
+    </form>
+
+
+Register the view in your admin class
+--------------------------------
```
[the remark this answers is not in the document]
Comment (jordygroote) [reply cookbook/add-tabs.rst:65]: Will fix

File: cookbook/add-tabs.rst (hunk 36 above)
[the remark this answers is not in the document]
Comment (jordygroote) [reply cookbook/add-tabs.rst:4]: Will try...

**Comment by jordygroote** [comment]
@danrot  @nnatter I think i covered all the feedback now.

**Comment by danrot** [comment]
Sorry, wanted to rebase that PR and accidentally closed it :see_no_evil: Gonna open a new one.

**Comment by danrot** [comment]
Will merge #559 as soon as tests succeeded. Thanks for your effort!

**Comment by jordygroote** [comment]
No problem! Happy 2 help.

**Comment by nnatter** [comment]
Thanks a lot @jordygroote!
